src/modules/iotjs_module_mqtt.c

```diff
@@ -40,46 +40,62 @@ JS_FUNCTION(MqttConstructor) {
       iotjs_jval_get_property(opts, IOTJS_MAGIC_STRING_USERNAME);
   jerry_value_t password =
       iotjs_jval_get_property(opts, IOTJS_MAGIC_STRING_PASSWORD);
-  jerry_value_t clientId =
+  jerry_value_t clientID =
       iotjs_jval_get_property(opts, IOTJS_MAGIC_STRING_CLIENT_ID);
+  jerry_value_t will = iotjs_jval_get_property(opts, IOTJS_MAGIC_STRING_WILL);
 
   MQTTPacket_connectData options = MQTTPacket_connectData_initializer;
   options.willFlag = 0;
   options.MQTTVersion = (int)iotjs_jval_as_number(version);
   options.keepAliveInterval = (int)iotjs_jval_as_number(keepalive);
   options.cleansession = 1;
 
-  if (jerry_value_is_string(username)) {
-    iotjs_string_t str = iotjs_jval_as_string(username);
-    MQTTString mqttstr = MQTTString_initializer;
-    mqttstr.lenstring.data = strdup((char*)iotjs_string_data(&str));
-    mqttstr.lenstring.len = (int)iotjs_string_size(&str);
-    options.username = mqttstr;
-    iotjs_string_destroy(&str);
-  }
-  if (jerry_value_is_string(password)) {
-    iotjs_string_t str = iotjs_jval_as_string(password);
-    MQTTString mqttstr = MQTTString_initializer;
-    mqttstr.lenstring.data = strdup((char*)iotjs_string_data(&str));
-    mqttstr.lenstring.len = (int)iotjs_string_size(&str);
-    options.password = mqttstr;
-    iotjs_string_destroy(&str);
-  }
-  if (jerry_value_is_string(clientId)) {
-    iotjs_string_t str = iotjs_jval_as_string(clientId);
-    MQTTString mqttstr = MQTTString_initializer;
-    mqttstr.lenstring.data = strdup((char*)iotjs_string_data(&str));
-    mqttstr.lenstring.len = (int)iotjs_string_size(&str);
-    options.clientID = mqttstr;
-    iotjs_string_destroy(&str);
+#define MQTT_OPTION_ASSIGN_FROM(owner, name)                           \
+  do {                                                                 \
+    if (jerry_value_is_string(name)) {                                 \
+      iotjs_string_t str = iotjs_jval_as_string(name);                 \
+      MQTTString mqttstr = MQTTString_initializer;                     \
+      mqttstr.lenstring.data = strdup((char*)iotjs_string_data(&str)); \
```

Comment:
data from `strdup` should be freed

Reply:
since strdup is terminated by '\0', binary data like payload cannot be fully copied, the message sent will be incomplete, copy via strndup is recommended

Reply:
May I ask you a github suggestion, please?